Python/compile.c

```diff
@@ -1655,16 +1655,10 @@ compiler_body(struct compiler *c, location loc, asdl_stmt_seq *stmts)
 static int
 compiler_codegen(struct compiler *c, mod_ty mod)
 {
-    _Py_DECLARE_STR(anon_module, "<module>");
-    RETURN_IF_ERROR(
-        compiler_enter_scope(c, &_Py_STR(anon_module), COMPILER_SCOPE_MODULE,
-                             mod, 1));
-
```

Comment:
What is the purpose of moving this from here out into the (only) two callers of this function? It duplicates this line of code, but it doesn't seem like it would make any other difference, AFAICT.

Reply:
The purpose is to be able to make a distinction between compiler_enter_scope failed, in which case we don't need to exit the scope, and the case where compiler_enter_scope succeeded but compiler_codegen failed later, in which case we do need to exit the scope. I don't think this is currently correct in all cases in main.

Reply:
That makes sense! Beyond adding the compilation modes option, I wasn't clear where this PR was aiming to maintain behavior vs fixing problems in existing behavior.

Looking at the definition of `compiler_exit_scope`, it seems like if we have one too many it would leave things in an observably broken state (like too small `c_nestlevel`?) Is there somewhere we can add asserts to show that things are broken today, but not after this fix?

Also, it seems like it would be simplest to get the enter/exit logic correct if we do it just once, inside `compiler_codegen`, rather than in both callers?

Reply:
I'm looking into it, I think it's a bigger mess than I previously thought. There are functions like compile_class which always exit the scope (which they did not enter). And VISIT_SEQ/VISIT_SEQ_IN_SCOPE also call compiler_exit_scope.

> Also, it seems like it would be simplest to get the enter/exit logic correct if we do it just once, inside `compiler_codegen`, rather than in both callers?

I don't think we can exit the scope in compiler_codegen, because we need it until after optimize_and_assemble.

Reply:
I'm going to revert the commit with the enter/exit changes for now - I probably only needed it in a broken state when I was sending bad inputs to the unit tests.